Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -500,22 +500,30 @@ class BrowserViewController: UIViewController {
             // This assumes that the DB returns rows in some kind of sane order.
             // It does in practice, so WFM.
             log.debug("Queue. Count: \(cursor.count).")
-            if cursor.count <= 0 {
-                return
-            }
+            if cursor.count > 0 {
+
+                // Filter out any tabs received by a push notification to prevent dupes.
+                let urls = cursor.flatMap { $0?.url.asURL }.filter { !receivedURLs.contains($0) }
+                if !urls.isEmpty {
+                    DispatchQueue.main.async {
+                        self.tabManager.addTabsForURLs(urls, zombie: false)
+                    }
+                }
 
-            // Filter out any tabs received by a push notification to prevent dupes.
-            let urls = Array(Set(cursor.flatMap { $0?.url.asURL }).subtracting(receivedURLs))
-            if !urls.isEmpty {
+                // Clear *after* making an attempt to open. We're making a bet that
+                // it's better to run the risk of perhaps opening twice on a crash,
+                // rather than losing data.
+                self.profile.queue.clearQueuedTabs()
+            }
+                
+            // Then, open any received URLs from push notifications.
+            if !receivedURLs.isEmpty {
                 DispatchQueue.main.async {
-                    self.tabManager.addTabsForURLs(urls, zombie: false)
+                    receivedURLs.forEach { receivedURL in
+                        self.switchToTabForURLOrOpen(receivedURL, isPrivileged: false)
```

Comment:
Does it make sense to keep switching to every tab in the "receivedURLs" list? Why not just switch directly to the last one. Fewer animations that way.